### ISSUE TITLE
SHAP vs LIME

### ISSUE THREAD

**Issue by andrewwlong** [shap/shap#19]
how would you explain how your algorithm (in xgboost) is different than LIME?

**Comment by slundberg** [comment]
Good and common question. LIME is a great method that was developed earlier here at UW in Carlos' group, and strongly influenced our work on SHAP. If you use the Shapley kernel we proposed as LIME's local kernel then the linear LIME method will produce estimates of SHAP values. So SHAP values come with the black box local estimation advantages of LIME, but also come with theoretical guarantees about consistency and local accuracy from game theory (attributes from other methods we unified).

A key limitation of the black box Kernel SHAP method above (and also of LIME) is that it needs to run many evaluations of the original model to estimate a single vector of feature importances. The Tree SHAP method is a fast algorithm specifically designed for tree ensemble models that can exactly compute the SHAP values in drastically less time.

So to summarize...SHAP values in XGBoost can be viewed as an improvement over the original linear LIME, and are computed exactly with a fast tree-specific algorithm.

**Comment by sudhanshunathmishra** [comment]
Hi, I had a quick question about the interpretation of the Shap values. In LIME, the locally weighted linear model is a regression. Therefore, one could interpret the coefficient for a feature X to be the change in the output for a unit change in X, holding the other features constant. Is a similar interpretation possible for the Shap value of feature X?

I am interested in using the local approximation to 'advise' how to change the binary classification of a point from 0 to 1 (or vice versa). I am able to do this using LIME but was wondering if it would be fair to use shap values for the same purpose.

**Comment by slundberg** [comment]
LIME is indeed a linear regression, but typically in a binary input space that may be different than the original input space (i.e. toggling super pixels). SHAP values can also be seen as regression coefficients just with a different local kernel (this is how Kernel SHAP works). Just remember these coefficients are also in a binary input space representing the impact of the feature taking it's current value vs. being unknown. So you could advise for example that points with high SHAP value magnitudes should move towards their mean feature value to lower the output.

As a broader comment, if you want to know how to best change the output, I would consider just running a simple optimizer like scipy.optimize.

**Comment by sudhanshunathmishra** [comment]
Thank you so much for the answers Scott. I had a couple of follow-ups:
1.
Let us say we have trained a model to predict the probability of default for a credit applicant.

i) Let a feature be 'number_of_prior_defaults' with mean value = 5
ii) Let the value of that feature for a particular applicant 'A' be 10
iii) Let Pr(default| 'A') = 0.65
iv) Let the shap value for 'number_of_prior_defaults' at 'A' be 0.30

One would expect that lower the value of 'number_of_prior_defaults', lower the probability of default (And vice versa). As you had mentioned, the shap value for this feature represents the impact of the feature taking its current value vs. being unknown (its mean). Therefore:

i) Is the best we can infer that applicant A would reduce his probability of default by moving 'number_of_prior_defaults' towards the mean value of 5?
ii) If he did change 'number_of_prior_defaults' to 5 (say vector A'), can we predict what his new  Pr(default|A') would be? 

I would guess that the probability would reduce by the shap value of 0.30. However, by this logic, a vector with each of its value at its feature's mean would result in Pr(default) = 0. This seems incorrect.

2.
If its not too much trouble, could you elaborate on how I might be able to use an optimizer for the task? To be concrete, I am interested to find the "closest" vector to an original vector that has a reverse prediction. The brute force way would be to search through the input space with increasing distance from the original vector. However as this seems computationally intensive,  I was hoping to use the local approximation regression coefficients in the original input space (analogous to a gradient) to predict sufficient changes required to reverse the prediction.

Thanks so much!

**Comment by sudhanshunathmishra** [comment]
Another question - is there a way to map the Shapley values (in the binary input space) to regression coefficients (in the original input space)? I am very curious to see how the regression coefficients would differ from those generated by LIME in the original input space for a particular prediction.

The most relevant information I found was in [this](https://mpra.ub.uni-muenchen.de/72116/1/MPRA_paper_72116.pdf) paper:

<img width="682" alt="screen shot 2018-07-07 at 6 09 41 pm" src="https://user-images.githubusercontent.com/10003061/42415043-1458d5a4-8211-11e8-8fd3-02728fbbe6a8.png">

**Comment by slundberg** [comment]
1. I don't think "However, by this logic, a vector with each of its value at its feature's mean would result in Pr(default) = 0" is correct. If all the features were at their expected values (and the model was linear), then Pr(default) = prior default rate, not 0.
2. There are many gradient free optimization methods (such as the Nelder-Mead simplex algorithm) that will do much better than brute force optimization. You can use SHAP values to help you figure out which features matter, and for the Tree SHAP this could be much faster. But for a model agnostic approach doing a direct optimization for the metric you care about will be faster.

3. If you take `shap_values / (X - X.mean(0))` you will get the regression coefficients if you are explaining a linear model. So you could use that as an approximation.

Overall for finding the minimal perturbations on a model I would compute the global feature importance with Tree SHAP then take the features that matter and put them into an optimizer like Nelder-Mead.

**Comment by sudhanshunathmishra** [comment]
Great! Thanks a lot, Scott. I just wanted to clarify point 3)

1. If I'm not mistaken, SHAP values can be generated for any model (using KernelShap). By "_you will get the regression coefficients if you are explaining a linear model_", do you mean the formula only holds for SHAP values computed for a linear model? For e.g., could we not apply the same technique to generate the linear regression approximation to a neural net at a particular observation?

2. You mentioned `shap_values / (X - X.mean(0))` gives the regression coefficients. Is this a regression with no offset or can that also be calculated?

**Comment by slundberg** [comment]
1. For a linear model the SHAP values are beta * (X - X.mean(0)). So you
can get a kind of linear approximation specific to a given sample by just
dividing by (X - X.mean(0)). This is only exact for linear models, for
non-linear models you can do it, but it is just an approximation that does
have any guarantees that I know about.
2. That would be with an offset (which would be the same as the shap_values
offset).

**Comment by bhrigs** [comment]
Thanks for the great Library. The only place I am struggling is to view the list of features with their boundaries with respect to a class for a prediction. The way we get the various thresholds/ranges in lime using `explainer.as_list()`. Is there something like this in Shap? Thanks!

![image](https://user-images.githubusercontent.com/19342526/84347050-8f347f00-ab7f-11ea-9866-77cd08fc24b3.png)
